File: 22. Generate Parentheses/22. Generate Parentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        def backtrack(open, close, current):
            if len(current) == 2 * n:
                result.append(current)
                return
            
            if open < n:
                backtrack(open + 1, close, current + '(')
            if close < open:
                backtrack(open, close + 1, current + ')')

        result = []
        backtrack(0, 0, '')
        return result

File: 22. Generate Parentheses/22. Generate Parentheses.cs
public class Solution {
    public IList<string> GenerateParenthesis(int n) {
        var result = new List<string>();
        Backtrack(result, new char[2 * n], 0, 0, n);
        return result;
    }
    
    private void Backtrack(List<string> result, char[] current, int open, int close, int max) {
        if (current.Length == open + close) {
            result.Add(new string(current));
            return;
        }
        
        if (open < max) {
            current[open + close] = '(';
            Backtrack(result, current, open + 1, close, max);
        }
        if (close < open) {
            current[open + close] = ')';
            Backtrack(result, current, open, close + 1, max);
        }
    }
}